dynamiclifecycle: Add DynamicLifecycle module

ovidiutirla opened the pull request:
Adds dynamic lifecycle module. A module which allows to group cells and manage their lifecycle as a unit. The dynamic lifecycle cell modules starts the lifecycle by leveraging  the dynamicconfig. The hooks are started only when the specific feature is enabled and also ensuring the dependencies are already started. 

```release-note
Allow to group cells lifecycle and control the enablement leveraging the dynamic-config.
```

ovidiutirla committed:
CODEOWNERS: Add dynamiclifecycle to sig-foundations

Signed-off-by: Ovidiu Tirla <otirla@google.com>

joamaki commented on pkg/featuremanager/cell.go:
```diff
@@ -0,0 +1,43 @@
+// see https://github.com/cilium/cilium/issues/34930
+var Cell = cell.Module(
+	"feature-manager",
+	"FeatureManager",
```
Describe a bit what this is instead of just repeating the id. E.g. if/when we generate documentation out of all modules it'd be nice to have something other than `feature-manager: FeatureManager` (not that I'm doing any better job with that...)

ovidiutirla replied:
The documentation part was omitted for the draft. But good suggestions, thanks.

joamaki commented on pkg/featuremanager/cell.go:
```diff
@@ -0,0 +1,43 @@
+)
+
+// Cell allows to manage the FeatureManager.
+// A feature is a group of cell hooks that are grouped together and their
```
I wonder if we could come up with a better name than "Feature" for these. even "DynamicFeature" would be better.

ovidiutirla replied:
renamed to DynamicFeature even though I feel like this naming is a bit awkward.

joamaki commented on pkg/featuremanager/cell.go:
```diff
@@ -0,0 +1,43 @@
+	"github.com/spf13/pflag"
+)
+
+// Cell allows to manage the FeatureManager.
```
Try to instead describe what this is providing. Whoever comes here wants to know what this is providing and doing and "allows to manage the FeatureManager" says absolutely nothing. E.g. try to describe the purpose, i.e. this manages dynamic lifecycles or something like that.

ovidiutirla replied:
Let me know if the current description make more sense.

joamaki commented on pkg/featuremanager/feature_manager.go:
```diff
@@ -0,0 +1,157 @@
+}
+
+type FeatureManager interface {
+	Append(feature Feature, hook cell.HookInterface)
```
Document these methods and also the fields in public structs.

ovidiutirla replied:
added

joamaki commented on pkg/featuremanager/feature_manager.go:
```diff
@@ -0,0 +1,157 @@
+	process(status []FeatureStatus, ctx context.Context) error
+}
+
+type DefaultFeatureManager struct {
```
Are we expecting to have many of these? If not, then maybe it's better to just have `type FeatureManager struct`, and maybe `type FeatureAppender interface` in addition for the cases where we want to be able to override/wrap it.

ovidiutirla replied:
I like having the interface as it defines a clear contract of a public provided type, including the ability to easily mocking.. I've refactored the code with the suggestions we've discussed, please have another look and let me know if you still want to remove the interface.

joamaki commented on pkg/featuremanager/feature_manager.go:
```diff
@@ -0,0 +1,157 @@
+		return
+	}
+
+	health.OK("OK")
```
Would be useful to report some stats, like how many features are running etc.

joamaki replied:
Oh and we probably want metrics too. Check `pkg/metrics/cell.go` and grep for `metrics.Metric` for examples.

ovidiutirla replied:
Let me tackle the metrics in a separate PR, let's agree on the design and we can improve the metrics after.

joamaki commented on pkg/featuremanager/feature_manager.go:
```diff
@@ -0,0 +1,157 @@
+	for _, feature := range status {
+		var err error
+		if feature.Enabled {
+			err = d.start(feature.Feature, ctx)
```
Shouldn't we use a context with a timeout here? Otherwise a misbehaving feature would prevent all others from being started. Though we might also need to account for `start` not caring about `ctx`.

ovidiutirla replied:
moved this logic to the stateDB reconciler.

joamaki commented on pkg/featuremanager/feature_manager.go:
```diff
@@ -0,0 +1,157 @@
+			case <-ctx.Done():
+				return ctx.Err()
+			case <-w:
+			default:
```
`default:` here will make this busy-loop!

also consider adding a rate-limiter to the loop just in case.

ovidiutirla replied:
sounds good, we'll do that.

joamaki commented on pkg/featuremanager/cell.go:
```diff
@@ -0,0 +1,43 @@
+
+func (c config) Flags(flags *pflag.FlagSet) {
+	flags.Bool("enable-feature-manager", c.EnableFeatureManager, "Enables support for dynamic feature management")
+	flags.String(FeatureStatusConfigKey, c.FeatureStatus, "List of features and their status")
```
Why is this a configuration flag? When would it be used? Why is it JSON?
I would also try to use a more descriptive flag name as `--feature-status` is a bit confusing.

ovidiutirla replied:
I want to keep the same standard of configuring cells. It's JSON to avoid having multiple keys as this includes quite a lot of configs. (we've done similar thing to CiliumEndpointSlice configuration). 

Also, we need this config as dynamically enabling would mean changing this which we'll be reflected by the dynamicconfig. But we should also allow this to work without the dynamicconfig cell by providing static values to this config.

Renamed the key to `DynamicFeatureConfig`, do you think it makes more sense now?

joamaki commented on pkg/featuremanager/feature_manager.go:
```diff
@@ -0,0 +1,157 @@
+
+	p.JobGroup.Add(job.OneShot("feature-manager-table-watcher", func(ctx context.Context, health cell.Health) error {
+		for {
+			entry, found, w := dynamicconfig.WatchKey(p.DB.ReadTxn(), p.Table, FeatureStatusConfigKey)
```
Hmm, this might be nicer if we pull this out into its own thing and break the direct dependency to `dynamicconfig`. E.g. depend on interface that gives the enabled features and a watch channel, and then implement that bridge between `dynamicconfig` and this in `cell.go` or even in `daemon/cmd/cells.go`.

Or an alternative way to do this is to have a separate table for controlling these dynamic features/lifecycles. The `Append` would insert the features into the table and we could then control that from e.g. `dynamicconfig` to turn them off and on. Nice thing about that would be it'd be easy to inspect it with cilium-dbg, and we could also have info there like when was it started and how long did the start take and stuff like that. Further that approach probably would make it easier to model dependencies between these in the future.

ovidiutirla replied:
resolving for now as I've moved the reconcile loop to stateDB reconciler.

ovidiutirla committed:
pkg/dynamiclifecycle: Add cell module

Adds the cell module to provide the watcher for dynamicconfig that allows to Add dynamic features,and a lifecycle wrapper that allows to append Hooks.
It delegates the reconciliation logic to a stateDB reconciler.

Signed-off-by: Ovidiu Tirla <otirla@google.com>

ovidiutirla committed:
pkg/dynamiclifecycle: Add DynamicLifecycle wrapper

Provides `WithDynamicLifecycle` which allows wrapping Cells into a DynamicFeature that would allow the Manager to manage its lifecycle.

Signed-off-by: Ovidiu Tirla <otirla@google.com>

joamaki commented on pkg/dynamiclifecycle/manager.go:
```diff
@@ -0,0 +1,207 @@
+// this specific feature can be enabled.
+// If the dependencies are not enabled the DynamicFeature will keep retrying until all dependencies are enabled.
+func (d *DefaultManager) AddFeature(feature DynamicFeature, deps ...DynamicFeature) {
+	_, _, found := d.tl.Get(d.db.ReadTxn(), ByFeature(feature))
```
I would do the lookup with the WriteTxn as otherwise this is racy.

ovidiutirla replied:
indeed.

joamaki commented on pkg/dynamiclifecycle/manager.go:
```diff
@@ -0,0 +1,207 @@
+// If one of the DynamicFeature hooks fail to start, the reconciler attempts
+// to stop already started hooks. The reconciler will retry to start the DynamicFeature.
+func (d *DefaultManager) Start(feature DynamicFeature) {
+	obj, _, found := d.tl.Get(d.db.ReadTxn(), ByFeature(feature))
```
Here as well I'd look up using WriteTxn and then `txn.Abort()` it if not found.

ovidiutirla replied:
👍

joamaki commented on pkg/dynamiclifecycle/manager.go:
```diff
@@ -0,0 +1,207 @@
+		return
+	}
+
+	obj.Enabled = true
```
We should perhaps check if it's already enabled and if so abort?

ovidiutirla replied:
I guess, not that necessary, but added for consistency.

joamaki commented on pkg/dynamiclifecycle/manager.go:
```diff
@@ -0,0 +1,207 @@
+// The Lifecycle hooks are stopped by the TableLifecycle reconciler.
+// Currently, without tracking the dependencies.
+func (d *DefaultManager) Stop(feature DynamicFeature) {
+	obj, _, found := d.tl.Get(d.db.ReadTxn(), ByFeature(feature))
```
Likewise, lookup with WriteTxn.

ovidiutirla replied:
yup

joamaki commented on pkg/dynamiclifecycle/manager.go:
```diff
@@ -0,0 +1,207 @@
+}
+
+// Append adds the specified hook to the list of hooks managed for the given DynamicFeature.
+func (d *DefaultManager) Append(feature DynamicFeature, hook cell.HookInterface) {
```
We should perhaps document a bit when `Append` can be called safely. I presume we'd want all the appends to happen before the application is started? We could also consider checking this invariant.

joamaki commented on pkg/dynamiclifecycle/manager.go:
```diff
@@ -0,0 +1,207 @@
+	// FeatureLifecycle.Append() is thread safe
+	obj.FeatureLifecycle.Append(hook)
+
+	d.l.Info("Appending hook to feature", "feature", feature)
```
This should be debug level logging?

ovidiutirla replied:
moved all logging to debugging. except the SetEnabled ones.

joamaki commented on pkg/dynamiclifecycle/table.go:
```diff
@@ -0,0 +1,80 @@
+// SetStatus sets the reconciliation status.
+// Used by the reconciler to update the reconciliation status of the memo.
+func (tl TableLifecycle) SetStatus(newStatus reconciler.Status) TableLifecycle {
+	println(newStatus.Kind)
```
remove debug printing

ovidiutirla replied:
😄

joamaki commented on pkg/dynamiclifecycle/table.go:
```diff
@@ -0,0 +1,80 @@
+	statusIndex = reconciler.NewStatusIndex(TableLifecycle.GetStatus)
+)
+
+type TableLifecycle struct {
```
Weird name for this. Could this be just called `DynamicFeature` and what was `DynamicFeature` could be `DynamicFeatureName`.

ovidiutirla replied:
renamed. also renamed the table name, which will be used in CLI.

joamaki commented on pkg/dynamiclifecycle/lifecycle.go:
```diff
@@ -0,0 +1,71 @@
+	IsRunning() bool
+}
+
+type DefaultLifecycle struct {
```
Do we expect multiple implementations of this? If not, I'd perhaps just call this `dynamicFeatureLifecycle`, or even just `lifecycle`.

joamaki replied:
Also how about just embedding all this into `TableLifecycle` instead? It does mean that we'll need a "lifecycle appender" that does database updates, but it should be simpler than having the table + mutating lifecycle objects. We'd also then have transactional semantics for all the state.

joamaki commented on pkg/dynamiclifecycle/lifecycle.go:
```diff
@@ -0,0 +1,71 @@
+	return string(f)
+}
+
+type DynamicFeatureLifecycle interface {
```
WDYT about calling this just `dynamiclifecycle.Lifecycle`?

joamaki commented on pkg/dynamiclifecycle/manager.go:
```diff
@@ -0,0 +1,207 @@
+	Enabled bool
+}
+
+type Manager interface {
```
Do we need the interface for this? Could just drop this and have `type Manager struct`.

ovidiutirla replied:
merged declaration up

joamaki commented on pkg/dynamiclifecycle/manager.go:
```diff
@@ -0,0 +1,207 @@
+// The Lifecycle hooks are started by the TableLifecycle reconciler considering the dependencies.
+// If one of the DynamicFeature hooks fail to start, the reconciler attempts
+// to stop already started hooks. The reconciler will retry to start the DynamicFeature.
+func (d *DefaultManager) Start(feature DynamicFeature) {
```
Since this just enables it, perhaps we could just have `SetEnabled(feature DynamicFeature, enabled bool)` instead of start and stop?

ovidiutirla replied:
💯

joamaki commented on pkg/dynamiclifecycle/reconciler.go:
```diff
@@ -0,0 +1,117 @@
+}
+
+func (o *ops) checkDependencies(dependencies []DynamicFeature, expectEnabled bool) error {
+	readTxn := o.db.ReadTxn()
```
Don't do a ReadTxn inside the ops as this is racy! Update is called with a ReadTxn so pass that in here. 

The reason it is racy is because Update() is called based on the state of status in a snapshot and it might've changed so we might read some future state here, which may or may not get us in trouble. Better to just use the one ReadTxn for everything to avoid issues.

ovidiutirla replied:
was thinking about the same thing, initially I though we could have deps changes so by having the ReadTxn from DB would get the fresh list of deps. quite hacky, anyway, moved to the provided ReadTxn and made the deps final and immutable

joamaki commented on pkg/dynamiclifecycle/reconciler.go:
```diff
@@ -0,0 +1,117 @@
+		if obj.Enabled != expectEnabled {
+			errs = errors.Join(errs, fmt.Errorf("expected DynamicFeature=%s to have Enabled=%t but was %t", dep, expectEnabled, obj.Enabled))
+			continue
+		}
```
Should we also check if it's running?

ovidiutirla replied:
added a check for this too.

joamaki commented on pkg/dynamiclifecycle/reconciler.go:
```diff
@@ -0,0 +1,117 @@
+// all dependants are running and if the requirement is met the lifecycle is started.
+func (o *ops) Update(ctx context.Context, txn statedb.ReadTxn, tl TableLifecycle) error {
+	// Nothing to do
+	if tl.Enabled == tl.FeatureLifecycle.IsRunning() {
```
Having mutating state in `FeatureLifecycle` makes me worried and it's harder to reason about this since it doesn't have the transactional semantics that `TableLifecycle` has. E.g. the reconciler wouldn't know if `IsRunning` changes and wouldn't know to call `Update` if we depend on it. I would store all state in `TableLifecycle` instead.

Note that the reconciler allows us to mutate the object. It does mean it'll need to be passed around by reference though (`*TableLifecycle`) and you'll need to explicitly shallow clone it then.

joamaki commented on pkg/dynamiclifecycle/reconciler.go:
```diff
@@ -0,0 +1,117 @@
+		return err
+	}
+
+	tl.StartedAt = o.clock.Now()
```
You'll need `*TableLifecycle` (passed by reference) to set these.

ovidiutirla commented on pkg/dynamiclifecycle/manager.go:
```diff
@@ -0,0 +1,174 @@
+	}
+
+	m.initializeDynamicFeaturesTable()
+	m.initDynamicConfigWatcher()
```
Hmm, maybe we need any synchronization for these two?

ovidiutirla replied:
added a waitForInit in the `initDynamicConfigWatcher`

ovidiutirla committed:
pkg/dynamiclifecycle: Register dynamiclifecycle

Adds the dynamiclifecycle.Cell to agent cells and register the table as an experimental table in CLI.

Signed-off-by: Ovidiu Tirla <otirla@google.com>

ovidiutirla committed:
docs: Include dynamic-lifecycles table docs

Signed-off-by: Ovidiu Tirla <otirla@google.com>

ovidiutirla commented on pkg/dynamiclifecycle/reconciler.go:
```diff
@@ -0,0 +1,133 @@
+}
+
+func getLifecycleForHooks(hooks []cell.HookInterface) cell.Lifecycle {
+	dl := cell.DefaultLifecycle{}
```
we need to change the implementation of `DefaultLifecycle`, as we currently we use this logic to construct the lifecycle for OnStop but the numStarted is not public so the lifecycle will think the hooks are not started.

ovidiutirla replied:
opened https://github.com/cilium/hive/pull/18

ovidiutirla commented on vendor/github.com/cilium/hive/cell/lifecycle.go:
```diff
@@ -75,6 +75,19 @@ type augmentedHook struct {
 	moduleID FullModuleID
 }
 
+func NewDefaultLifecycle(hooks []HookInterface, numStarted int, logThreshold time.Duration) *DefaultLifecycle {
```
commit to be dropped before merging

joamaki commented on pkg/dynamiclifecycle/manager.go:
```diff
@@ -0,0 +1,174 @@
+		return
+	}
+	if obj.Enabled == enabled {
+		d.l.Warn("Nothing to do", "feature", feature, "enabled", enabled)
```
Is the warning really necessary? Wouldn't we hit this often when the configmap is changed?

ovidiutirla replied:
We shouldn't hit this often as we only watch a specific key, though you are quite right, removed the log.

joamaki commented on pkg/dynamiclifecycle/manager.go:
```diff
@@ -0,0 +1,174 @@
+// If enabled is marked false the Lifecycle hooks are stopped by the DynamicFeature reconciler.
+// Currently, without tracking the dependencies.
+func (d *manager) setEnabled(feature DynamicFeatureName, enabled bool) {
+	writeTxn := d.db.WriteTxn(d.dft)
```
This creates a write transaction per feature which is a bit wasteful. Consider refactoring this to do a WriteTxn in the dynamic config watcher and use it also for the reading, e.g. the whole processing round happens within a single database transaction.

ovidiutirla replied:
makes sense

joamaki commented on pkg/dynamiclifecycle/manager.go:
```diff
@@ -0,0 +1,174 @@
+}
+
+func (d *manager) initDynamicConfigWatcher() {
+	d.jg.Add(job.OneShot("dynamic-config-watcher", func(ctx context.Context, health cell.Health) error {
```
I would move the `Add` call into `register*` and have `func (m *manager) dynamicConfigWatcher(ctx context.Context, health cell.Health) error`.

ovidiutirla replied:
Done

joamaki commented on pkg/dynamiclifecycle/manager.go:
```diff
@@ -0,0 +1,174 @@
+}
+
+func provideFeatureManager(p params) manager {
+	return manager{
```
Consider just embedding `params` into `manager` instead of repeating the fields.
Also, there's no need to "provide" the manager. I would also in general avoid "managers"
and instead break things up by functionality in to "small agents that do one thing"

I personally would just do something like:
```
type configWatcher struct {
  params
  limiter *rate.Limiter
}
func (w *configWatcher) watch(ctx context.Context, health cell.Health) error {
  ...
}

func registerWatcher(p params) error {
  w := &configWatcher{p, rate.NewLimiter(time.Second, 1)}
  p.JobGroup.Add(job.OneShot("dynamic-config-watcher", w.watch))
}

func initializeTable(p params) error {
  ...
}

var Cell = cell.Module(...,
  cell.Invoke(
    initializeTable,
    registerWatcher,
  ),
)
```

ovidiutirla replied:
moved type params struct into cell.go, do you see a better place to fit that piece?

joamaki commented on pkg/dynamiclifecycle/table.go:
```diff
@@ -0,0 +1,121 @@
+
+	ByFeature = featureIndex.Query
+
+	statusIndex = reconciler.NewStatusIndex((*DynamicFeature).getStatus)
```
Is this needed?

joamaki commented on pkg/dynamiclifecycle/manager.go:
```diff
@@ -0,0 +1,174 @@
+		jg:  p.JobGroup,
+		mc:  p.ManagerConfig,
+		fr:  p.FeatureRegistrations,
+		rt:  rate.NewLimiter(1*time.Second, 1),
```
Put this into a const at the top.

joamaki commented on pkg/dynamiclifecycle/reconciler.go:
```diff
@@ -0,0 +1,133 @@
+
+func getLifecycleForHooks(hooks []cell.HookInterface, running bool) cell.Lifecycle {
+	if running {
+		return cell.NewDefaultLifecycle(hooks, len(hooks), 100*time.Millisecond)
```
Put the log thresholds into consts at the top.

joamaki reviewed (changes requested):
Starting to look good!

joamaki commented on pkg/dynamiclifecycle/table.go:
```diff
@@ -0,0 +1,153 @@
+		"IsRunning",
+		"StartedAt",
+		"StoppedAt",
+		"StartDuration",
```
Also add the `Status` to this.

ovidiutirla replied:
added `Status`

joamaki commented on pkg/dynamiclifecycle/watcher.go:
```diff
@@ -0,0 +1,155 @@
+}
+
+func waitForDynamicFeatureTableInit(ctx context.Context, p watcherParams) error {
+	//Wait for DynamicFeatureTable initialization
```
Why do we need this?

ovidiutirla replied:
no longer required as we moved the table init in the provide(). removed

joamaki commented on pkg/dynamiclifecycle/cell.go:
```diff
@@ -0,0 +1,54 @@
+// The manager uses the DynamicConfig table as the source of truth for enablement.
+// The manager delegates the responsibility of enablement to DynamicFeature StateDB reconciler.
+var Cell = cell.Module(
+	"dynamic-feature-manager",
```
"dynamic-lifecycle" to match package name?

ovidiutirla replied:
renamed to `"dynamic-lifecycle-manager"`

joamaki commented on pkg/dynamiclifecycle/reconciler.go:
```diff
@@ -0,0 +1,137 @@
+)
+
+var (
+	logThreshold = 100 * time.Millisecond
```
Hmm i think it'd be fine to have `0` logging threshold for the dynamically started lifecycles

ovidiutirla replied:
changed to `0 * time.Millisecond`

joamaki commented on pkg/dynamiclifecycle/watcher.go:
```diff
@@ -0,0 +1,155 @@
+}
+
+func registerWatcher(p watcherParams) error {
+	if !p.ManagerConfig.EnableDynamicFeatureManager || !p.DynamicConfigCellConfig.EnableDynamicConfig {
```
Should we actually error out if `EnableDynamicFeatureManager` (`EnableDynamicLifecycles`?) is enabled but `EnableDynamicConfig` is disabled? Now this will silently not work if misconfigured.

ovidiutirla replied:
added a error log if DynamicLifecycleManager is enabled and DynamicConfig is disabled

joamaki replied:
I would still prefer erroring out. Error logs are easy to miss.

ovidiutirla replied:
changed to return error

joamaki reviewed (commented):
Starting to look really good!
